### ISSUE TITLE
Binding of Self within nested functions

### ISSUE THREAD

**Issue by Glyphack** [astral-sh/ty#1173]
### Summary

In the following example when we annotate self with type var in f then f binds that type var and the `Self` in b is actually bound by `f`:

```py
from typing import Self

class C[T]():
    def f(self: Self):
        def b(x: Self):
            reveal_type(x)  # revealed: Self@f
```

If we annotate `self` with `C`:

```py
from typing import Self

class C():
    def f(self: "C"):
        def b(x: Self):
            reveal_type(x)  # revealed: Self@b
```

And the first example is correct based on [this todo comment](https://github.com/astral-sh/ruff/blob/59c8fda3f8f3bf2cc5c1ae34e7ca9dbea4d0278f/crates/ty_python_semantic/resources/mdtest/annotations/self.md#L33) in mdtests.s

**Comment by Glyphack** [comment]
I want to work on this issue.

**Comment by Glyphack** [comment]
One interesting result if we bind self to the current typevar_binding_context is this diagnostic:

```py
from typing import Self

class Shape:
    def nested_func(self: Self) -> Self:
        def inner() -> Self:
            reveal_type(self)  # [invalid-return-type] "Return type does not match returned value: expected `Self@inner`, found `Self@nested_func`"
            return self
        return inner()
```

Pyright and Mypy have different behaviors here:
- https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbABQ5AxiQIYDO9hAFrQkQFzlQ9QAmRYFBRF6MInwD6wAK4oqACnqlgHQioCUUALQA%2BdWS69j-QZhQiQCrXoOruJxyCIA3IrRKT47JZp4BiKGc3Dwk1YjIAARExCWk5KgdHY2cYGRAUKGUyJJNU9MzUS2tyIA
- https://gist.github.com/mypy-play/766aa5e340a00e8a27e4d6f69b476be1
